examples_tests/60.ClusteredRendering/cull_common.glsl

```diff
@@ -0,0 +1,247 @@
+	cone.baseRadius = cone.height * tanOuterHalfAngle;
+
+	return cone;
+}
+
+// Todo(achal): Rename to getClusterAABB
+nbl_glsl_shapes_AABB_t getCluster(in uvec3 localClusterID, in vec3 levelMinVertex, in float voxelSideLength)
+{
+	const vec3 camPos = pc.camPosGenesisVoxelExtent.xyz;
+
+	nbl_glsl_shapes_AABB_t cluster;
+	cluster.minVx = levelMinVertex + (localClusterID * voxelSideLength) + camPos;
+	cluster.maxVx = levelMinVertex + (localClusterID + uvec3(1u)) * voxelSideLength + camPos;
+	return cluster;
+}
+
+#define PLANE_COUNT 6
+vec4[PLANE_COUNT] getAABBPlanes(in nbl_glsl_shapes_AABB_t aabb)
+{
+	vec4 planes[PLANE_COUNT];
+
+	// 157
+	vec3 p0 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.minVx.z);
+	vec3 p1 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.maxVx.z);
+	vec3 p2 = vec3(aabb.maxVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	planes[0].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[0].w = dot(planes[0].xyz, p0);
+
+	// 013
+	p0 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.minVx.z);
+	p1 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.minVx.z);
+	p2 = vec3(aabb.maxVx.x, aabb.maxVx.y, aabb.minVx.z);
+	planes[1].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[1].w = dot(planes[1].xyz, p0);
+
+	// 402
+	p0 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.minVx.z);
+	p2 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.minVx.z);
+	planes[2].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[2].w = dot(planes[2].xyz, p0);
+
+	// 546
+	p0 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p2 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	planes[3].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[3].w = dot(planes[3].xyz, p0);
+
+	// 451
+	p0 = vec3(aabb.minVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.maxVx.z);
+	p2 = vec3(aabb.maxVx.x, aabb.minVx.y, aabb.minVx.z);
+	planes[4].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[4].w = dot(planes[4].xyz, p0);
+
+	// 762
+	p0 = vec3(aabb.maxVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	p1 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.maxVx.z);
+	p2 = vec3(aabb.minVx.x, aabb.maxVx.y, aabb.minVx.z);
+	planes[5].xyz = normalize(cross(p1 - p0, p2 - p0));
+	planes[5].w = dot(planes[5].xyz, p0);
+
+	return planes;
+}
+
+bvec3 and(in bvec3 a, in bvec3 b)
+{
+	return bvec3(a.x && b.x, a.y && b.y, a.z && b.z);
+}
+
+bvec3 or(in bvec3 a, in bvec3 b)
+{
+	return bvec3(a.x || b.x, a.y || b.y, a.z || b.z);
+}
+
+float projectedSphericalVertex(in vec3 origin, in vec3 planeNormal, in vec3 pos)
+{
+	return dot(normalize(pos - origin), planeNormal);
+}
+
+vec3 findFarthestPointOnConeInDirection(in vec3 planeNormal, in cone_t cone)
+{
+	const vec3 m = cross(cross(planeNormal, cone.direction), cone.direction);
+	const vec3 farthestBasePoint = cone.tip + (cone.direction * cone.height) - (m * cone.baseRadius); // farthest to plane's surface, away from positive half-space
+	return farthestBasePoint;
+}
```

Comment:
make it a "method" of `nbl_glsl_shapes_cone_t` and put it in `nbl/builtin/glsl/shapes/cone.glsl`

```glsl
vec3 nbl_glsl_shapes_cone_t_findFarthestPointInDirection(in cone_t cone, in vec3 dir)
```